### PR TITLE
Enable ghost zones > 1 for octrees

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

This enables support for more than 1 ghost zones with octree datasets.

Also add a minor speed improvement of the mesh-sampling implementation (always start with the largest domain).

## PR Checklist

- [x] pass `black --check yt/`
- [x] pass `isort . --check --diff`
- [x] pass `flake8 yt/`
- [x] pass `flynt yt/ --fail-on-change --dry-run -e yt/extern`
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by cphyc** [comment]
Note that using more than one ghost zone is very slow, but it works nonetheless.

File: yt/frontends/ramses/data_structures.py [hunk 1]
```diff
@@ -362,23 +365,28 @@ def retrieve_ghost_zones(self, ngz, fields, smoothed=False):
...
+            new_subset = _subset_with_gz[ngz]
             mylog.debug(
-                "Reusing previous subset with ghost zone for domain %s", self.domain_id
+                "Reusing previous subset with %s ghost zones for domain %s",
+                ngz,
+                self.domain_id,
             )
-        except AttributeError:
+        except KeyError:
             new_subset = RAMSESDomainSubset(
```
Comment (matthewturk) [inline yt/frontends/ramses/data_structures.py:378]: Would this result in, for instance, having both the entire subset *and* the subset + GZ in memory?

**Review by matthewturk — APPROVED** [review]

File: yt/frontends/ramses/data_structures.py (hunk 1 above)
Comment (cphyc) [reply yt/frontends/ramses/data_structures.py:378]: Yes; the reason I decided to store them in memory is that computing the ghost zones is (very) slow, as it grows roughly as `N log(N) (Ngz**3-8)`, where `N` is the number of octs in the subset.

**Comment by matthewturk** [comment]
Makes sense, thanks!

**Pull request merged**